torch/_jit_internal.py

```diff
@@ -216,6 +229,43 @@ def __getattr__(self, key):
 
     return createResolutionCallbackFromEnv(closure_lookup())
 
+def check_can_compile_named_tuple(cls, loc):
+    """
+    Check that a NamedTuple can be compiled. This means that it must not have additional
+    user-defined methods.
+
+    Arguments:
+        cls: The NamedTuple subclass in question.
+        loc: A SourceRange instance corresponding to the definition of cls.
+    """
+    # These are methods common to all NamedTuples. Using an allowlist is not the cleanest approach,
+    # but these methods don't exist on NamedTuple or tuple, so it is not possible to write code like:
+    #   for name, value in cls.__dict__.items():
+    #       if getattr(NamedTuple, name) is None and getattr(tuple, name) is None:
+    #           # Must be a user-defined function.
+    allowed_methods = [
+        "__annotations__",
+        "__doc__",
+        "__getnewargs__",
+        "__module__",
+        "__new__",
+        "__repr__",
+        "__slots__",
+        "_asdict",
+        "_field_defaults",
+        "_field_types",
+        "_fields",
+        "_make",
+        "_replace",
+        "_source",
+    ]
+
+    for name, value in cls.__dict__.items():
+        if name not in allowed_methods and inspect.isfunction(value):
+            raise FrontendError(loc, "User-defined methods in NamedTuples are currently not supported")
```

Comment:
Even this is not quite what we want - the `loc` here is the source location of the use of the name of the `NamedTuple` class that caused it to be compiled, not the source code for it. So, the `loc` doesn't highlight the name or source of the problematic method.